File: 2020/haskell/src/DockingData.hs
{-# LANGUAGE NamedFieldPuns #-}

module DockingData where

import qualified Data.Map.Strict as M
import           Data.List        (isPrefixOf)
import           AdventOfCode
import           Data.Char        (isDigit)
import           Data.Bits

data Line = Mask { andMask :: Int, orMask  :: Int, maskStr :: String }
          | Instr { address :: Int, value :: Int }
          deriving (Show)

parseBinary ::  Bool -> String -> Int
parseBinary rep = bin2Dec . map translate
  where translate '0' = False
        translate '1' = True
        translate 'X' = rep

parse :: String -> Line
parse line
  | "mask" `isPrefixOf` line = parseMask line
  | otherwise                = parseInstr line
  where parseMask line = let maskStr = drop 7 line in
                           Mask (parseBinary True maskStr) (parseBinary False maskStr) maskStr
        parseInstr line = let (addr:_:value:[]) = words line in
                            Instr (read $ takeWhile (\c -> isDigit c) $ drop 4 addr) (read value)

solve14a :: [Line] -> Int
solve14a program = M.foldl (+) 0 $ go M.empty (0, 0) program
  where go :: M.Map Int Int -> (Int, Int) -> [Line] -> M.Map Int Int
        go memory _     []      = memory
        go memory masks (x:xs) = case x of
                                   Mask a o _       -> go memory (a,o) xs
                                   Instr addr value -> go (M.insert addr (applyMasks masks value) memory) masks xs
        applyMasks (andMask, orMask) value = (value .|. orMask) .&. andMask

floatingAddresses :: String -> Int -> [Int]
floatingAddresses maskStr pos = go maskStr pos []
  where
    go :: String -> Int -> [Int] -> [Int]
    go []      _   acc = acc ++ [0]
    go maskStr pos acc
          | (last maskStr) == '0' = acc ++ (map (\a -> 2 * a + pos `mod` 2) $ go (subs maskStr) (pos `div` 2) acc)
          | (last maskStr) == '1' = acc ++ (map (\a -> 2 * a + 1) $ go (subs maskStr) (pos `div` 2) acc)
          | (last maskStr) == 'X' = acc ++ (concat (map (\a -> [2 * a + 0, 2 * a + 1]) $ go (subs maskStr) (pos `div` 2) acc))
    subs s = substring 0 (pred $ length s) s

solve14b :: [Line] -> Int
solve14b program = M.foldl (+) 0 $ go M.empty "" program
  where go :: M.Map Int Int -> String -> [Line] -> M.Map Int Int
        go memory _ [] =  memory
        go memory mask (x:xs) = case x of
                                  Mask  _ _  ms    -> go memory ms xs
                                  Instr addr value -> go (insertAll memory $ (zip (floatingAddresses mask addr) (repeat value))) mask xs
        insertAll :: M.Map Int Int -> [(Int, Int)] -> M.Map Int Int
        insertAll m xs = M.union (M.fromList xs) m

day14 :: [String] -> IO ()
day14 input = do
  let program = map parse input
  print $ solve14a program
  print $ solve14b program

File: 2020/rust/src/dockingdata.rs
use std::collections::HashMap;

fn solve14a(program: Vec<&str>) -> u64 {

    let mut memory: HashMap<u64, u64> = HashMap::new();

    let mut and_mask: u64 = 0;
    let mut or_mask: u64 = 0;

    for line in program {
        if line.starts_with("mask") {

            let mut mask = line.clone().trim().get(7..).unwrap().replace("X", "0");
            or_mask = u64::from_str_radix(&mask, 2).unwrap();

            mask = line.trim().get(7..).unwrap().replace("X", "1");
            and_mask = u64::from_str_radix(&mask, 2).unwrap();
        } else {
            let parts: Vec<&str> = line.split(" = ").collect();
            let address: u64 = parts[0][4..parts[0].len() - 1].parse::<u64>().unwrap();
            let mut value = parts[1].parse::<u64>().unwrap();

            value = (value | or_mask) & and_mask;
            memory.insert(address, value);
        }
    }

    memory.values().sum()

}

fn floating_addresses(pos: u64, mask: String) -> Vec<u64> {

    let mut addresses: Vec<u64> = Vec::new();

    if mask.len() == 0 {
        addresses.push(0);
        return addresses;
    }

    if mask.chars().last().unwrap() == '0' {
        for addr in floating_addresses(pos / 2, mask[0..mask.len()-1].to_string()) {
            addresses.push(2 * addr + pos % 2);
        }
    } else if mask.chars().last().unwrap() == '1' {
        for addr in floating_addresses(pos / 2, mask[0..mask.len()-1].to_string()) {
            addresses.push(2 * addr + 1);
        }
    } else if mask.chars().last().unwrap() == 'X' {
        for addr in floating_addresses(pos / 2, mask[0..mask.len()-1].to_string()) {
            addresses.push(2 * addr + 0);
            addresses.push(2 * addr + 1);
        }
    } else {
        println!("blah: {}", mask.chars().last().unwrap());
    }

    addresses
}

fn solve14b(program: Vec<&str>) -> u64 {

    let mut memory: HashMap<u64, u64> = HashMap::new();

    let mut mask: String = String::from("");

    for line in program {
        if line.starts_with("mask") {
            mask = line.get(7..).unwrap().trim().to_string().clone();
        } else {
            let parts: Vec<&str> = line.split(" = ").collect();
            let address: u64 = parts[0][4..parts[0].len() - 1].parse::<u64>().unwrap();
            let value = parts[1].parse::<u64>().unwrap();

            let addresses = floating_addresses(address, mask.clone());
            for addr in addresses {
                memory.insert(addr, value);
            }
        }
    }

    memory.values().sum()
}

// Day 14: Docking Data
pub fn solve(data: String) {

    let program: Vec<&str> = data.split('\n').collect();

    let part1 = solve14a(program.clone());
    println!("Part 1: {}", part1);

    let part2 = solve14b(program);
    println!("Part 2: {}", part2);
}


#[cfg(test)]
mod tests {

    use super::*;
    use std::fs;

    fn load_input_file(filename: &str) -> String {
        fs::read_to_string(filename).unwrap().trim().to_string()
    }

    #[test]
    fn testinput_part1() {
        let input = load_input_file("../inputs/testinput14a.txt");
        let program: Vec<&str> = input.split('\n').collect();
        assert_eq!(solve14a(program), 165);
    }

    #[test]
    fn testinput_part2() {
        let input = load_input_file("../inputs/testinput14b.txt");
        let program: Vec<&str> = input.split('\n').collect();
        assert_eq!(solve14b(program), 208);
    }

    #[test]
    fn solution() {
        let input = load_input_file("../inputs/input14.txt");
        let program: Vec<&str> = input.split('\n').collect();
        assert_eq!(solve14a(program.clone()), 14925946402938);
        assert_eq!(solve14b(program), 3706820676200);
    }

}
